### ISSUE TITLE
PHP Fatal error: Interface 't3lib_Singleton' not found

### ISSUE THREAD

**Issue by fritzdacat** [franzholz/tt_products#22]
After changing a TYPO3 7 installation to composer, I always get this error in the Pagemodul if a tt_products Contentelement is there. In ListModule it works.

PHP Fatal error: Interface 't3lib_Singleton' not found in /Volumes/Webs/htdocs/xxxx/composer/web/typo3conf/ext/div2007/hooks/class.tx_div2007_hooks_cms.php on line 41

TYPO3 7.6.23
tt_products 2.8.19
div2007 1.7.13 (loaded as dependency of products)

**Comment by franzholz** [comment]
This is a problem of your autoloader cache. The Install Tools offers a "Dump Autoload Information".

**Comment by fritzdacat** [comment]
I know only composer dump-autoload. The TYPO3 Install Tool does not have this function.

If I composer dump-autoload make and also with Composer update come this error:

Warning: Ambiguous class resolution, "t3lib_div" found in both "/Volumes/Webs/htdocs/xx/composer/web/typo3conf/ext/div2007/Migrations/Code/LegacyClassesForIde.php" and "/ Volumes / Webs / htdocs / xx / composer / web / typo3conf / ext / div2007 / class.tx_div2007_compatibility6.php ", the first will be used.
Warning: Ambiguous class resolution, "tslib_cObj" found in both "/Volumes/Webs/htdocs/xx/composer/web/typo3conf/ext/div2007/Migrations/Code/LegacyClassesForIde.php" and "/ Volumes / Webs / htdocs / xx / composer / web / typo3conf / ext / div2007 / class.tx_div2007_compatibility6.php ", the first will be used.
Warning: Ambiguous class resolution, "t3lib_extMgm" what found in both "/Volumes/Webs/htdocs/xx/composer/web/typo3conf/ext/div2007/Migrations/Code/LegacyClassesForIde.php" and "/ Volumes / Webs / htdocs / xx / composer / web / typo3conf / ext / div2007 / class.tx_div2007_compatibility6.php ", the first will be used.
Warning: Ambiguous class resolution, "tslib_fe" what found in both "/Volumes/Webs/htdocs/xx/composer/web/typo3conf/ext/div2007/Migrations/Code/LegacyClassesForIde.php" and "/ Volumes / Webs / htdocs / xx / composer / web / typo3conf / ext / div2007 / class.tx_div2007_compatibility6.php ", the first will be used.
Warning: Ambiguous class resolution, "t3lib_parsehtml" what found in both "/Volumes/Webs/htdocs/xxx/composer/web/typo3conf/ext/div2007/Migrations/Code/LegacyClassesForIde.php" and "/ Volumes / Webs / htdocs / xx / composer / web / typo3conf / ext / div2007 / class.tx_div2007_compatibility6.php ", the first will be used.

**Comment by franzholz** [comment]
The file ext / div2007 / class.tx_div2007_compatibility6.php is only included at all by tt_products if the class t3lib_div does not exist. However it must already exist from the Migrations folder. class.tx_div2007_core.php has the function activateCompatibility6. This case should never happen. 
The call for activateCompatibility6 has been removed in tt_products 2.8.20 already, because all TYPO3 4.5 class names are available by the Migrations tool.

**Comment by fritzdacat** [comment]
Thank you for your reply. This refers however only to my 2 answer. and tt_products 2.8.20 is not yet officially released.

My main problem is: PHP Fatal error: Interface 't3lib_Singleton' not found in /Volumes/Webs/htdocs/xxxx/composer/web/typo3conf/ext/div2007/hooks/class.tx_div2007_hooks_cms.php on line 41

**Comment by franzholz** [comment]
The interface  't3lib_Singleton'  is defined in the Migrations folder of div2007. If your TYPO3 environment does not find it, then the reason are the caches of the autoloader. You must clear the caches in the Install Tool.

**Comment by fritzdacat** [comment]
I deleted both caches. And also compiled the Web once again with Composer. Unfortunately, the problem still exists.

**Comment by franzholz** [comment]
All the TYPO3 4.5 classes are loaded by the TYPO3 class alias loader.
 see https://github.com/TYPO3/class-alias-loader

In your case it still uses some caches from a former div2007 version.
div2007/Migrations/Code/LegacyClassesForIde.php does not initialize the class, because otherwise the classes would be duplicate. ClassAliasMapResult.php is run by the TYPO3 class alias loader. You can see that the classes are all there.

**Comment by fritzdacat** [comment]
But what can I do, except delete all the caches of the project. What is funny, is the Web without Composer works

**Comment by franzholz** [comment]
Does the Composer need anything more in the file composer.json?
Maybe the classmap must include the Migrations folder.

**Comment by fritzdacat** [comment]
Please excuse me, but what do you mean "Maybe the classmap must include the migrations folder."?

**Comment by franzholz** [comment]
I do not understand what you have done in TYPO3 with the composer. What does "change the TYPO3 installation to composer" mean? The composer is used by TYPO3 as a standard.

**Comment by fritzdacat** [comment]
Wir übernehmen gerade eine TYPO3 7 Installation, welche ohne Composer gemacht wurde. 

Diese will ich gerade auf composer umbauen.

Alles funktioniert auch wunderbar.  

But always get this error in the Pagemodul if a tt_products Contentelement is there. In ListModule it works.

PHP Fatal error: Interface 't3lib_Singleton' not found in /Volumes/Webs/htdocs/xxxx/composer/web/typo3conf/ext/div2007/hooks/class.tx_div2007_hooks_cms.php on line 41.

This is my Composer.json:

{
  "name": "typo3",
  "repositories": [
    {
      "type": "composer",
      "url": "https://composer.typo3.org/"
    }
  ],
  "minimum-stability": "dev",
  "prefer-stable": true,
  "require": {
    "typo3/cms": "7.6.23",
    "typo3-ter/tt-products": "2.8.19",
    "typo3-ter/addons-tt-products": "0.4.2",
    "typo3-ter/bootstrap-core": "1.2.8",
    "typo3-ter/bootstrap-grids": "1.3.1",
    "typo3-ter/bsdist": "1.3.3",
    "typo3-ter/metaseo": "2.0.1",
    "typo3-ter/powermail": "3.9.0",
    "typo3-ter/realurl": "2.1.5",
    "typo3-ter/vhs": "3.1.0",
    "typo3-ter/sr-feuser-register": "4.0.2",
    "typo3-ter/static-info-tables": "6.3.10" 
    },
  "autoload": {
    "psr-4": {
    }
  },
  "extra": {
    "typo3/cms": {
      "cms-package-dir": "{$vendor-dir}/typo3/cms",
      "web-dir": "web"
    }
  },
  "config": {
    "secure-http": false,
    "bin-dir": "."
  }
}

**Comment by fritzdacat** [comment]
Maybe we can also talk via Slack. Would also be willing to pay something if you help me with the problem.

**Comment by LimeUwe** [comment]
We encountered the same problem.
@fritzdacat did you manage to solve it somehow?
(extensions are all new newest available)

**Comment by franzholz** [comment]
div2007 must also be added to the composer.json. And since div2007 2.10.0 the extension migration_core must be used.

**Comment by LimeUwe** [comment]
@franzholz 
I tried as suggested in the project I need it the updated version, but it still doesn't work.
For a test, I installed the tt_products extension via composer in a fairly clean typo3 "bootstrap" distribution. All the depencies were managed cleanly, so all required depencies, "table","div2007","core_migrare"..., were installed too. After adding the typoscripts to the template I placed a product list plugin on a page and opened the site - the result was still the same error message from above.

May it be, that the "core_migration" extension does not work in composer based installations?

**Comment by franzholz** [comment]
I have never found any TYPO3 environment where the extension "core_migration" did not work. This extension works fine with every composer environment on which I did test it.
Any problems have always turned out to be a cache which inhibited to use the new extension files.
You did not mention which versions of which extensions you are using.

**Comment by LimeUwe** [comment]
Hi Franz,

I required typo3-ter/tt_products (2.8.22) in the main composer.json file of a plain TYPO3 8 with bootstrap distribution. The composer loaded following dependencies:

Package operations: 5 installs, 0 updates, 0 removals
  - Installing typo3-ter/tsparser (0.3.0): Loading from cache
  - Installing typo3-ter/div2007 (1.10.6): Loading from cache
  - Installing typo3-ter/table (0.7.0): Loading from cache
  - Installing typo3-ter/migration-core (0.1.2): Loading from cache
  - Installing typo3-ter/tt-products (2.8.22): Loading from cache

I found out, that the alias map from "mirgration_core" was not correctly merged by composer.

We had to put the alias map configuration from the migration_core composer file 

`"extra": {
		"typo3/class-alias-loader": {
	        "class-alias-maps": [
	          "web/typo3conf/ext/migration_core/Migrations/Code/ClassAliasMap.php"
	        ]
	    }
	},`

to our main project composer file. After that, the shop extension started working.

Best regards,
Uwe

**Comment by franzholz** [comment]
Could you try to copy the contents of migration_core/Migrations/Code/ClassAliasMapResult.php into the file migration_core/Migrations/Code/ClassAliasMap.php instead of the "require_once 'ClassAliasMapResult.php';" statement? Maybe the require_once statement does not work correctly.

**Comment by typecat** [comment]
I've encountered the same problem with an installation via composer using:
* TYPO3 8.7.18
* EXT:tt_products 2.8.22
* EXT:div2007 1.9.0
* EXT:migration_core 0.1.3

My colleague and I analysed the autoload files created by composer, dumped the caches multiple times, but no luck.
There is one file that may cause the problem:
in EXT:div2007/Migrations/Code/LegacyClassesForIde.php in the first lines:
```
if (class_exists('TYPO3\\ClassAliasLoader\\ClassAliasMapGenerator')) {
    return array();
} else {
```
As described in the PHP manual (http://php.net/manual/en/function.class-exists.php)
>This function checks whether or not the given class has been defined. 
So maybe what happens here is, that the method returns true, because the class actually _is_ defined, but simply not loaded in composer mode. Then an empy array is returned instead of the class alias maps .

**Comment by franzholz** [comment]
This is an older version of div2007. Newer versions of div2007 do not have a Migrations folder any more.
And you must not install an older version of div2007 together with migration_core because both have the same migration mapping classes.

**Comment by typecat** [comment]
I've required the latest version of EXT:div2007. Still the same problem.

**Comment by franzholz** [comment]
Then you did not get the correct div2007. There is no Migrations folder!

**Comment by typecat** [comment]
I didn't mean, that there is a Migrations folder. I meant, that I still get the error. Only this time, I have no clue what may be the cause. Using EXT:div2007 1.10.8 now.

But I've discovered something strange: if I install EXT:migration_core it via`composer require typo3-ter/migration-core ^0.1`, then I get the error `PHP Fatal error: Interface 't3lib_Singleton' not found`. 

BUT: If I download it manually form TER and then install it as a local package via `composer require jambagecom/migration-core dev-master` (dev-master here, because it's a local package in my project git), then the error disappears, because `'t3lib_singleton' => 'TYPO3\\CMS\\Core\\SingletonInterface'` is now registered in vendor/composer/autoload_classaliasmap.php . In the previous method it wasn't.

I've never seen such behaviour before and don't know what the cause could be.

**Comment by franzholz** [comment]
You shoul better ask here and give a feedback:

https://github.com/TYPO3/class-alias-loader

**Comment by franzholz** [comment]
migration_core is not needed any more with the latest published tt_products 2.9.7
